Scoped extension building for C++ backed custom ops tests

xmfan opened the pull request:
FIXES #125579 #131103 #133197 #133283 #134738 #135369 #135685

Tests that create C++ extensions can cause flakiness in CI due to library namespace conflict and test ordering. We can build them in temp dirs to ensure isolation.

An alternative is to build these as part of the build process and have build time errors.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #136695



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov @rec

xmfan committed:
scoped extension building for C++ backed custom ops tests

[ghstack-poisoned]

xmfan committed:
Update on "Scoped extension building for C++ backed custom ops tests"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

xmfan committed:
Update on "Scoped extension building for C++ backed custom ops tests"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

xmfan committed:
Update on "Scoped extension building for C++ backed custom ops tests"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

xmfan committed:
Update on "Scoped extension building for C++ backed custom ops tests"

FIXES #125579 #131103 #133197 #133283 #134738 #135369 #135685




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

xmfan committed:
Update on "Scoped extension building for C++ backed custom ops tests"

FIXES #125579 #131103 #133197 #133283 #134738 #135369 #135685




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

xmfan committed:
Update on "Scoped extension building for C++ backed custom ops tests"

FIXES #125579 #131103 #133197 #133283 #134738 #135369 #135685




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

xmfan committed:
Update on "Scoped extension building for C++ backed custom ops tests"

FIXES #125579 #131103 #133197 #133283 #134738 #135369 #135685




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

zou3519 commented on test/dynamo/test_misc.py:
```diff
@@ -79,6 +79,7 @@
 from torch.testing._internal.common_utils import (
     freeze_rng_state,
     IS_FBCODE,
+    scoped_load_inline,
```
@xmfan can you give more context about what this PR is (and put it into the PR body?)

zou3519 replied:
It sounds like this PR is supposed to fix flakiness, do you know what caused the flakyness?

zou3519 commented on torch/testing/_internal/common_utils.py:
```diff
@@ -5365,11 +5366,20 @@ def remove_cpp_extensions_build_root():
         else:
             shutil.rmtree(default_build_root, ignore_errors=True)
+
+def scoped_load_inline(*args, **kwargs):
+    temp_dir = tempfile.TemporaryDirectory()
+    if kwargs.get("verbose", False):
+        print(f'Using temporary extension directory {temp_dir.name}...', file=sys.stderr)
+    assert "build_directory" not in kwargs
+    kwargs["build_directory"] = temp_dir.name
+    module = cpp_extension.load_inline(*args, **kwargs)
+    return module, temp_dir
```
It's only really scoped if it's a context manager. Otherwise it's just creating a C++ extension with a temporary directory. The options are:
1) change this API to be a context manager. On exit, this will delete the temp_dir.
2) rename the API to be "c++ extension with temporary directory, where the directory will *eventually* clean itself up, we don't know when"

First one is generally more pythonic

zou3519 reviewed (commented):
I don't think the naming makes sense, also left some questions

xmfan committed:
Update on "Scoped extension building for C++ backed custom ops tests"

FIXES #125579 #131103 #133197 #133283 #134738 #135369 #135685

Tests that create C++ extensions can cause flakiness in CI due to library namespace conflict and test ordering. An alternative is to build these as part of the build process.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

zou3519 commented on test/test_autograd.py:
```diff
@@ -9899,19 +9900,19 @@ def backward(ctx, gO):
 }
 
-TORCH_LIBRARY(test_autograd_cpp_node, m) {
+TORCH_LIBRARY(test_multigrad_all_hooks, m) {
```
Why did this need to change?

xmfan commented:
@pytorchbot merge

xmfan committed:
Update on "Scoped extension building for C++ backed custom ops tests"

FIXES #125579 #131103 #133197 #133283 #134738 #135369 #135685

Tests that create C++ extensions can cause flakiness in CI due to library namespace conflict and test ordering. An alternative is to build these as part of the build process.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov rec

[ghstack-poisoned]

xmfan commented:
@pytorchbot merge

xmfan committed:
Update on "Scoped extension building for C++ backed custom ops tests"

FIXES #125579 #131103 #133197 #133283 #134738 #135369 #135685

Tests that create C++ extensions can cause flakiness in CI due to library namespace conflict and test ordering. We can build them in temp dirs to ensure isolation.

An alternative is to build these as part of the build process and have build time errors.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov rec

[ghstack-poisoned]

xmfan committed:
Update on "Scoped extension building for C++ backed custom ops tests"

FIXES #125579 #131103 #133197 #133283 #134738 #135369 #135685

Tests that create C++ extensions can cause flakiness in CI due to library namespace conflict and test ordering. We can build them in temp dirs to ensure isolation.

An alternative is to build these as part of the build process and have build time errors.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov rec

[ghstack-poisoned]

xmfan committed:
Update on "Scoped extension building for C++ backed custom ops tests"

FIXES #125579 #131103 #133197 #133283 #134738 #135369 #135685

Tests that create C++ extensions can cause flakiness in CI due to library namespace conflict and test ordering. We can build them in temp dirs to ensure isolation.

An alternative is to build these as part of the build process and have build time errors.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov rec

[ghstack-poisoned]

xmfan commented:
@pytorchbot merge

xmfan commented:
@pytorchbot merge
